Add function byteSwap

Priyansh121096 opened the pull request:
Fixes #54734.

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Added function `byteSwap` which reverses the bytes of unsigned integers. This is particularly useful for reversing values of types which are represented as unsigned integers internally such as IPv4.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

Priyansh121096 committed:
Add function `byteSwap`.

byteSwap accepts an integer `operand` and returns the integer which is
obtained by swapping the **endianness** of `operand` i.e. reversing the
bytes of the `operand`.

Issue: #54734

Priyansh121096 committed:
Implement byteswap for following dtypes.

UInt[8|16|32|64]

TODOs:
- Improve NOT_IMPLEMENTED error message.
- Add implementation for FixedStrings (reverse the bytes).
- See whether this needs to be implemented for UInt[128|256] and
  signed integers as well.

Priyansh121096 committed:
Docs and comment

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=55211) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=55211) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=55211) it.</sub>

Priyansh121096 commented:
Hello, I've added an implementation for the `byteSwap` function asked for in #54734. The example given in the issue works with this implementation:
```result
┌─toIPv4(3351772109)─┐
│ 199.199.251.205    │
└────────────────────┘

┌─toIPv4(byteSwap(3351772109))─┐
│ 205.251.199.199              │
└──────────────────────────────┘
```

I have the following queries with regards to this feature:
1. I have currently implemented `byteSwap` for `UInt[8|16|32|64]` types.
  a. Do we want to implement it for `UInt[128|256]` as well? There's no `__builtin_bswap[128|256]` available in clang but we can add our custom implementation if required.
  b. Do we want to implement it for any other data types (`Int*`, `FixedString`, etc.)?
2. As per [this](https://github.com/ClickHouse/ClickHouse/blob/14767fd4ac081cea93bae725c25d8f775f1e7d55/src/Functions/FunctionsConversion.h#L237), conversion between numeric types and IPv6 is not allowed. This means that even though `byteSwap` can work with `UInt128` (which is how IPv6s are represented internally - [ref](https://clickhouse.com/docs/en/sql-reference/data-types/ipv6)), one cannot use the results of `byteSwap` to reverse IPv6s in the same way as one can with IPv4s.
3. Can this function be considered "injective"? IIUC, if a function is injective, no two distinct inputs lead to the same output. While that seems to be the case here, I'm confused as to whether we consider a function injective across data types or not. For example, `byteSwap(0xFF00) == 0xFF == byteSwap(0xFF)`. I would think that since the two byteSwaps produce values of different data types (`UInt16` and `UInt8` respectively), they can be considered distinct even though their "value" is the same. Could someone please advise?

Priyansh121096 committed:
Add tests

Priyansh121096 committed:
Remove comment

Priyansh121096 commented on src/Functions/byteSwap.cpp:
```diff
@@ -0,0 +1,79 @@
+template <typename T>
+inline T byteSwap(T)
+{
+    throw Exception(ErrorCodes::NOT_IMPLEMENTED, "byteSwap() is not implemented for {} datatype", demangle(typeid(T).name()));
```
Right now for anything greater than or equal to `2^64`, this prints the following error:
```
SELECT byteSwap(18446744073709552000.)

Query id: 7e644036-8f5a-42ac-be5e-ee1f2286c46e


0 rows in set. Elapsed: 0.005 sec. 

Received exception from server (version 23.9.1):
Code: 48. DB::Exception: Received from localhost:9000. DB::Exception: byteSwap() is not implemented for double datatype: While processing byteSwap(18446744073709552000.). (NOT_IMPLEMENTED)
```

I'd love to know if there's a way to get better type names.

Priyansh121096 replied:
Here's my attempt at an implementation for UInt128:
```
template <typename T>
requires std::is_same_v<T, UInt128>
inline T byteSwap(T x)
{
    UInt64 lower_half = x & 0xFFFFFFFFFFFFFFFF;
    UInt64 upper_half = (x >> 64) & 0xFFFFFFFFFFFFFFFF;
    UInt64 swapped_lower_half = __builtin_bswap64(lower_half);
    UInt64 swapped_upper_half = __builtin_bswap64(upper_half);
    UInt128 new_upper_half = static_cast<UInt128> (swapped_lower_half) << 64;
    UInt128 new_lower_half = static_cast<UInt128> (swapped_upper_half);
    return new_upper_half | new_lower_half;
}
```
But I'm not able to test the same. As soon as I go over (2^64-1), for some reason, it doesn't reach the implementation. Also notice that the last four digits of the input change (1616 -> 2000.). Any reasons why this could be happening?
```
clickhouse-400817.internal :) SELECT byteSwap(18446744073709551616);

SELECT byteSwap(18446744073709552000.)

Query id: 0743873e-f48b-423f-a160-1b160f5b25c3


0 rows in set. Elapsed: 0.125 sec. 

Received exception from server (version 23.9.1):
Code: 48. DB::Exception: Received from localhost:9000. DB::Exception: byteSwap() is not implemented for double datatype: While processing byteSwap(18446744073709552000.). (NOT_IMPLEMENTED)
```

rschu1ze replied:
> But I'm not able to test the same. As soon as I go over (2^64-1), for some reason, it doesn't reach the implementation. Also notice that the last four digits of the input change (1616 -> 2000.). Any reasons why this could be happening?

Because the example is interpreted as float:
```sql
SELECT toTypeName(18446744073709551616);

┌─toTypeName(18446744073709552000.)─┐
│ Float64                           │
└───────────────────────────────────┘
```

Try: `SELECT byteSwap(18446744073709551616::UInt128);`

I wonder why byteSwap on UInt64 and UInt128 operates on 2x8, respectively 4x8 bytes instead of 1x16 / 1x32 bytes? To swap all bytes uniformly, you could use `reverseMemcpy()` (base/base/unaligned.h).

alexey-milovidov commented on src/Functions/byteSwap.cpp:
```diff
@@ -0,0 +1,79 @@
+namespace
+{
+template <typename T>
+requires std::is_same_v<T, UInt8>
```
By the way, there is `std::byteswap` that works for all supported data types - let's use it to simplify.

Priyansh121096 replied:
Isn't that only available from C++23 onwards?

alexey-milovidov replied:
Yes, we use C++23.

Priyansh121096 replied:
Oh, awesome! This might be a stupid question but how exactly does one figure out which version of C++ a project is compatible with at any given point (I'm quite new to C++ development)?

Priyansh121096 replied:
I'll use `std::byteswap`. FYI, it even accepts signed integers. This means that people will be able to use byteswap with negative integers as well. Example:
```
┌─byteSwap(-255)─┐
│            511 │
└────────────────┘
```

Priyansh121096 replied:
Ah, makes sense. Thanks!

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #55211 from Priyansh121096/ClickHouse:master -->
*This is an automated comment for commit ed2c85bbce36360fe1a4cf4590937281aa6aaae5 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55211/ed2c85bbce36360fe1a4cf4590937281aa6aaae5/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6435772161">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55211/ed2c85bbce36360fe1a4cf4590937281aa6aaae5/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

alexey-milovidov commented on tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference:
```diff
@@ -192,6 +192,7 @@ blockSerializedSize
 blockSize
 buildId
 byteSize
+byteSwap
```
All new functions must be documented inline in the code, see `asin.cpp`.

The markdown docs are secondary to the documentation in code, they can contain a copy-paste.

Priyansh121096 replied:
Done.

alexey-milovidov commented:
> Do we want to implement it for UInt[128|256] as well?

Let's try.

> Can this function be considered "injective"? IIUC, if a function is injective, no two distinct inputs lead to the same output. While that seems to be the case here, I'm confused as to whether we consider a function injective across data types or not.

It's ok to make this function considered injective. The injectiveness property is used to eliminate function application, e.g. from GROUP BY. It means that only the domain of one data type is relevant.

Priyansh121096 committed:
Address a few review comments.

- Consider byteswap injective.
- Make function case-insensitive.
- Add in-code documentation and copy-paste it to the markdown docs.

Priyansh121096 committed:
Use std::byteswap instead of custom implementations.

- Also allow signed ints now because std::byteswap accepts them.
- Fix for style check.

Priyansh121096 commented:
> > Do we want to implement it for UInt[128|256] as well?
> 
> Let's try.

See https://github.com/ClickHouse/ClickHouse/pull/55211#discussion_r1349586579.
 
> > Can this function be considered "injective"? IIUC, if a function is injective, no two distinct inputs lead to the same output. While that seems to be the case here, I'm confused as to whether we consider a function injective across data types or not.
> 
> It's ok to make this function considered injective. The injectiveness property is used to eliminate function application, e.g. from GROUP BY. It means that only the domain of one data type is relevant.

This is done. 

I've also made this case-insensitive so that people can use both `byteSwap` and `byteswap`.

Priyansh121096 committed:
byteSwap is now documented; remove it from the list of undocumented functions

rschu1ze commented on docs/en/sql-reference/functions/math-functions.md:
```diff
@@ -711,3 +711,41 @@ Result:
+
+## byteSwap
+
+Accepts an unsigned integer `operand` and returns the integer which is obtained by swapping the **endianness** of `operand` i.e. reversing the bytes of the `operand`.
```
~unsigned~

Priyansh121096 replied:
Done.

rschu1ze commented on docs/en/sql-reference/functions/math-functions.md:
````diff
@@ -711,3 +711,41 @@ Result:
+
+One use-case of this function is reversing IPv4s:
+```result
+┌─toIPv4(3351772109)─┐
````
You probably want to include an example where an IPv4 is casted to an integer, which is then `byteSwap`ed, and casted to an IPv4 again.

Priyansh121096 replied:
> Do you mean that instead of the two statements, we can just demonstrate reversing IPv4s with a single statement like the following

Done

rschu1ze commented on tests/queries/0_stateless/02887_byteswap.sql:
```diff
@@ -0,0 +1,34 @@
+SELECT byteSwap(0);
```
Negative tests would be nice, e.g. what happens when there is > 1 argument, what happens when the inputs are of type != 8-64-integers (e.g. `Strings`s). Expected errors can be checked with the `serverError` syntax (grep for it in tests/queries/0_stateless/).

Priyansh121096 replied:
Added.

Priyansh121096 committed:
Remove "unsigned" from docs now that we support all ints.

Priyansh121096 committed:
Add negative tests.

Also:
- Add comments in tests.
- Add an example in docs where an IPv4 is casted to an int, byteswapped
  and then casted back to an IPv4.

rschu1ze committed:
Some fixups

Priyansh121096 commented on src/Functions/byteSwap.cpp:
```diff
@@ -0,0 +1,97 @@
+2. Reverse the bytes, i.e. C7 C7 FB CD -> CD FB C7 C7
+3. Convert the result back to an integer assuming big-endian, i.e. CD FB C7 C7  -> 3455829959
+
+One use-case of this function is reversing IPv5s:
```
@rschu1ze , shouldn't this say `IPv4s` and not `IPv5s` or am I missing something?

```suggestion
One use-case of this function is reversing IPv4s:
```

rschu1ze replied:
That was a typo, thanks ;-)

Priyansh121096 commented on src/Functions/byteSwap.cpp:
```diff
@@ -0,0 +1,97 @@
+                {"32-bit", "SELECT byteSwap(3351772109)", "3455829959"},
+                {"64-bit", "SELECT byteSwap(123294967295)", "18439412204227788800"},
+            },
+            .categories{"Mathematical"}},
```
@rschu1ze I see that you've moved the markdown docs to `arithmetic-functions.md`. Should we do the following?
```suggestion
            .categories{"Mathematical", "Arithmetic"}},
```

rschu1ze replied:
Thanks, the categories are currently not used for much but let's do as you propose.

rschu1ze committed:
Update src/Functions/byteSwap.cpp

Co-authored-by: Priyansh Agrawal <agrawal.priyansh@yahoo.in>

rschu1ze committed:
Update src/Functions/byteSwap.cpp

Co-authored-by: Priyansh Agrawal <agrawal.priyansh@yahoo.in>

Priyansh121096 commented:
@rschu1ze @alexey-milovidov, thanks for the merge! Hoping to make many more contributions in the future.
